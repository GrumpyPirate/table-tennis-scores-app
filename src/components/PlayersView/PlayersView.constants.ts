import { rem } from 'polished';
import styled from 'styled-components';

import styleConfig from 'config/styles';
import Button from 'components/Button/Button';

export const CreatePlayer = styled.div`
  @media ${styleConfig.mediaQueries.mobileOnly} {
    ${Button} {
      display: block;
      width: 100%;
    }
  }
`;

export const PlayersList = styled.ul`
  ${styleConfig.mixins.listReset}

  @media ${styleConfig.mediaQueries.tablet} {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-column-gap: ${rem(16)};
  }
`;
