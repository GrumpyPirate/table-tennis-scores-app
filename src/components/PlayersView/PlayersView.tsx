import React, { FunctionComponent, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import Button from 'components/Button/Button';
import PlayerListingItem from 'components/PlayerListingItem/PlayerListingItem';
import TabHeader from 'components/TabHeader/TabHeader';
import { AppDispatch, RootState } from 'store';
import { createPlayer } from 'store/slices/players';

import { CreatePlayer, PlayersList } from './PlayersView.constants';

const PlayersView: FunctionComponent = () => {
  const dispatch = useDispatch<AppDispatch>();
  const { allPlayers, isFetching } = useSelector((state: RootState) => ({
    allPlayers: state.players.allPlayers,
    isFetching: state.players.isFetching,
    error: state.players.error,
  }));

  const onCreatePlayerClick = useCallback(() => {
    const playerName = window.prompt('Enter player name:');

    if (playerName) {
      dispatch(createPlayer(playerName));
    }
  }, [dispatch]);

  return (
    <>
      <TabHeader title="Current Players" showSpinner={isFetching}>
        <CreatePlayer>
          <Button type="button" onClick={onCreatePlayerClick} disabled={isFetching}>
            Create Player
          </Button>
        </CreatePlayer>
      </TabHeader>
      <PlayersList>
        {allPlayers.map((player) => (
          <li key={player.id}>
            <PlayerListingItem player={player} />
          </li>
        ))}
      </PlayersList>
    </>
  );
};

export default PlayersView;
