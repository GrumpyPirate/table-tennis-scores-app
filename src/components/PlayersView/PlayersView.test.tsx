import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';

import createMockStore from 'testing/utils';

import PlayersView from './PlayersView';

describe('Components', () => {
  describe('PlayersView', () => {
    const mockStore = createMockStore();
    const windowPromptSpy = jest.fn();

    beforeEach(() => {
      jest.spyOn(window, 'prompt').mockImplementation(windowPromptSpy);
    });

    afterEach(() => {
      windowPromptSpy.mockClear();
    });

    test('should render as expected, without crashing', () => {
      const { container } = render(
        <Provider store={mockStore}>
          <PlayersView />
        </Provider>,
      );

      expect(container).toMatchSnapshot();
    });

    test('should prompt for a player name, when clicking the "Create Player" button', () => {
      render(
        <Provider store={mockStore}>
          <PlayersView />
        </Provider>,
      );

      expect(windowPromptSpy).not.toHaveBeenCalled();

      const button = screen.getByText('Create Player');
      fireEvent.click(button);

      expect(windowPromptSpy).toHaveBeenCalled();
    });
  });
});
