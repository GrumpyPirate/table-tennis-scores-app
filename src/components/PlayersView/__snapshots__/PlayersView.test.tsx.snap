// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components PlayersView should render as expected, without crashing 1`] = `
<div>
  <header
    class="sc-bqyKva dDkKpL"
  >
    <div
      class="sc-crrsfI lmciGZ"
    >
      <h1
        class="sc-dQppl bgPUqp"
      >
        Current Players
      </h1>
      <figure
        class="sc-iqHYGH iAWWDc"
      >
        <svg>
          loader.svg
        </svg>
      </figure>
    </div>
    <div
      class="sc-kstrdz fMSFqt"
    >
      <button
        class="sc-bdfBwQ gxmLa-d"
        type="button"
      >
        Create Player
      </button>
    </div>
  </header>
  <ul
    class="sc-hBEYos fMsFPQ"
  />
</div>
`;
